server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -646,8 +646,6 @@ protected BlobContainer blobContainer() {
      * Public for testing.
      */
     public BlobStore blobStore() {
-        assertSnapshotOrGenericThread();
```

Comment:
Looks like there are cases when the store can get lazily initialised. Lets ascertain that there are no real remote store interactions and this is just setting the references. code for lazy initialisation - https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java#L660

Reply:
Digging through the [Google](https://github.com/opensearch-project/OpenSearch/blob/681e5548c11eab389e19a75e3cf3053fb85fdcd4/plugins/repository-gcs/src/main/java/org/opensearch/repositories/gcs/GoogleCloudStorageRepository.java#L126-L129), [Azure](https://github.com/opensearch-project/OpenSearch/blob/681e5548c11eab389e19a75e3cf3053fb85fdcd4/plugins/repository-azure/src/main/java/org/opensearch/repositories/azure/AzureRepository.java#L166-L181), and [S3](https://github.com/opensearch-project/OpenSearch/blob/ec3473708326fcb126f6a4440bbda93f4a1a6c14/plugins/repository-s3/src/main/java/org/opensearch/repositories/s3/S3Repository.java#L326-L329) implementations, it does appear that they all just initialize fields and do not make any remote calls. The [HDFS implementation](https://github.com/opensearch-project/OpenSearch/blob/681e5548c11eab389e19a75e3cf3053fb85fdcd4/plugins/repository-hdfs/src/main/java/org/opensearch/repositories/hdfs/HdfsRepository.java#L124) however is much more complicated and I'm honestly not sure what it is doing.

Reply:
Thanks @andrross for checking. I will get HDFS implementation checked and update on this comment thread.

Reply:
HDFS definitely makes call to the configured namenodes as part of `createBlockStore`. 

Got following exception on passing the host where HDFS is not running:

Reply:
I still think this is a safe change as `blobStore()` is invoked while creating the repository and subsequent runs would be a no-op. So, the index creation flow which is a cluster state update will not be blocked due to this change.

@ashking94 @andrross What do you think?

Reply:
I think I'm okay with assertion removal. If you look at the [commit where the assertion was added](https://github.com/opensearch-project/OpenSearch/commit/b1bf643e41b2bf2bc1360887b4768836c9159214), the ultimate goal was "It makes possible to register e.g. s3repo with bucket that is not available at the moment." and I don't think we're going to change anything related to that. @ashking94 What do you think?

Reply:
IMO, we should ensure that removal of this check does not put regression on any critical flows like cluster state publication. If we remove the check, any future change which brings [this code](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java#L660) in critical workflows can lead to catastrophic behaviour of the cluster. If a repository is taking some time for bootstrapping, then it can lead to node getting removed from cluster. 

@sachinpkale  if we are aware of `store == null` not being true for our use cases, then may be we can have a dedicated threadpool and add assert on `store != null`.

Reply:
Adding assert on `store == null` would fail on node reboot. That means, first index creation post node reboot (assuming repository is already created) will calls `createBlobStore` implementation of the respective repository. This would be expensive in case of HDFS.

IMO, repository implementation should make sure not to make remote calls as part of `createBloblStore` (as done in S3, Azure and GCP implementation of the repository).

Reply:
Created a tracking issue: https://github.com/opensearch-project/OpenSearch/issues/6007